In [24]:
import pandas as pd

In [25]:
df = pd.read_parquet('../data/yellow_tripdata_2025-01.parquet')

df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,1,2025-01-01 00:18:38,2025-01-01 00:26:59,1.0,1.60,1.0,N,229,237,1,10.0,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0.0
1,1,2025-01-01 00:32:40,2025-01-01 00:35:13,1.0,0.50,1.0,N,236,237,1,5.1,3.5,0.5,2.02,0.0,1.0,12.12,2.5,0.0,0.0
2,1,2025-01-01 00:44:04,2025-01-01 00:46:01,1.0,0.60,1.0,N,141,141,1,5.1,3.5,0.5,2.00,0.0,1.0,12.10,2.5,0.0,0.0
3,2,2025-01-01 00:14:27,2025-01-01 00:20:01,3.0,0.52,1.0,N,244,244,2,7.2,1.0,0.5,0.00,0.0,1.0,9.70,0.0,0.0,0.0
4,2,2025-01-01 00:21:34,2025-01-01 00:25:06,3.0,0.66,1.0,N,244,116,2,5.8,1.0,0.5,0.00,0.0,1.0,8.30,0.0,0.0,0.0


In [26]:
print(f"Dataset Shape: {df.shape}")
df.info()
df.head()

Dataset Shape: (3475226, 20)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3475226 entries, 0 to 3475225
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64      

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,1,2025-01-01 00:18:38,2025-01-01 00:26:59,1.0,1.60,1.0,N,229,237,1,10.0,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0.0
1,1,2025-01-01 00:32:40,2025-01-01 00:35:13,1.0,0.50,1.0,N,236,237,1,5.1,3.5,0.5,2.02,0.0,1.0,12.12,2.5,0.0,0.0
2,1,2025-01-01 00:44:04,2025-01-01 00:46:01,1.0,0.60,1.0,N,141,141,1,5.1,3.5,0.5,2.00,0.0,1.0,12.10,2.5,0.0,0.0
3,2,2025-01-01 00:14:27,2025-01-01 00:20:01,3.0,0.52,1.0,N,244,244,2,7.2,1.0,0.5,0.00,0.0,1.0,9.70,0.0,0.0,0.0
4,2,2025-01-01 00:21:34,2025-01-01 00:25:06,3.0,0.66,1.0,N,244,116,2,5.8,1.0,0.5,0.00,0.0,1.0,8.30,0.0,0.0,0.0


In [27]:
# 1. Dataset Overview and Column Descriptions

# Column meanings and their relevance for hourly demand prediction:

In [28]:
# Display column descriptions and key information
import json

# Load the schema to understand columns
with open('../yellow_taxi_schema.json', 'r') as f:
    schema = json.load(f)

# Create a mapping of column names to descriptions
column_descriptions = {field['field_name']: field.get('description', 'No description') 
                       for field in schema['fields']}

print("=" * 80)
print("DATASET COLUMNS AND DESCRIPTIONS")
print("=" * 80)
for col, desc in column_descriptions.items():
    print(f"\n{col}: {desc}")

print("\n" + "=" * 80)
print(f"Total columns in dataset: {df.shape[1]}")
print(f"Total rows: {df.shape[0]}")
print("=" * 80)

DATASET COLUMNS AND DESCRIPTIONS

VendorID: A code indicating the TPEP provider that provided the record.

tpep_pickup_datetime: The date and time when the meter was engaged.

tpep_dropoff_datetime: The date and time when the meter was disengaged.

passenger_count: The number of passengers in the vehicle.

trip_distance: The elapsed trip distance in miles reported by the taximeter.

RatecodeID: The final rate code in effect at the end of the trip.

store_and_fwd_flag: Indicates whether the trip record was held in vehicle memory before sending (store and forward).

PULocationID: TLC Taxi Zone in which the taximeter was engaged.

DOLocationID: TLC Taxi Zone in which the taximeter was disengaged.

payment_type: A numeric code signifying how the passenger paid for the trip.

fare_amount: The time-and-distance fare calculated by the meter.

extra: Miscellaneous extras and surcharges.

mta_tax: Tax that is automatically triggered based on the metered rate in use.

tip_amount: Tip amount - au

# 2. Data Quality Assessment

Identifying and detecting potential data quality issues:

In [29]:
# Check for missing values
print("\n" + "=" * 80)
print("MISSING VALUES")
print("=" * 80)
missing_df = pd.DataFrame({
    'Column': df.columns,
    'Missing_Count': df.isnull().sum(),
    'Missing_Percent': (df.isnull().sum() / len(df)) * 100
})
missing_df = missing_df[missing_df['Missing_Count'] > 0].sort_values('Missing_Percent', ascending=False)
if len(missing_df) > 0:
    print(missing_df.to_string())
else:
    print("No missing values found!")

# Check for duplicate records
print("\n" + "=" * 80)
print("DUPLICATE RECORDS")
print("=" * 80)
duplicates = df.duplicated().sum()
print(f"Number of exact duplicate rows: {duplicates}")
print(f"Percentage of dataset: {(duplicates / len(df)) * 100:.2f}%")

# Check data types
print("\n" + "=" * 80)
print("DATA TYPES")
print("=" * 80)
print(df.dtypes)

# Check for potential data quality issues in numeric columns
print("\n" + "=" * 80)
print("NUMERIC COLUMNS - POTENTIAL ANOMALIES")
print("=" * 80)
numeric_cols = df.select_dtypes(include=['number']).columns

for col in numeric_cols:
    zero_count = (df[col] == 0).sum()
    negative_count = (df[col] < 0).sum()
    
    if zero_count > 0 or negative_count > 0:
        print(f"\n{col}:")
        if zero_count > 0:
            print(f"  - Zero values: {zero_count} ({(zero_count/len(df))*100:.2f}%)")
        if negative_count > 0:
            print(f"  - Negative values: {negative_count} ({(negative_count/len(df))*100:.2f}%)")

# Check date ranges
print("\n" + "=" * 80)
print("DATE/TIME RANGES")
print("=" * 80)
print(f"Pickup datetime range: {df['tpep_pickup_datetime'].min()} to {df['tpep_pickup_datetime'].max()}")
print(f"Dropoff datetime range: {df['tpep_dropoff_datetime'].min()} to {df['tpep_dropoff_datetime'].max()}")

# Check invalid trip durations
df['trip_duration_minutes'] = (df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60
invalid_trips = (df['trip_duration_minutes'] <= 0).sum()
print(f"Trips with invalid duration (dropoff before pickup): {invalid_trips}")

# Check location IDs
print("\n" + "=" * 80)
print("LOCATION IDS")
print("=" * 80)
print(f"Unique pickup zones: {df['PULocationID'].nunique()}")
print(f"Unique dropoff zones: {df['DOLocationID'].nunique()}")
print(f"Location ID range: {df['PULocationID'].min()} to {df['PULocationID'].max()}")
print(f"Unknown/null locations (PU): {(df['PULocationID'] == 0).sum()}")
print(f"Unknown/null locations (DO): {(df['DOLocationID'] == 0).sum()}")


MISSING VALUES
                                    Column  Missing_Count  Missing_Percent
passenger_count            passenger_count         540149        15.542845
RatecodeID                      RatecodeID         540149        15.542845
store_and_fwd_flag      store_and_fwd_flag         540149        15.542845
congestion_surcharge  congestion_surcharge         540149        15.542845
Airport_fee                    Airport_fee         540149        15.542845

DUPLICATE RECORDS
Number of exact duplicate rows: 0
Percentage of dataset: 0.00%

DATA TYPES
VendorID                          int32
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag               object
PULocationID                      int32
DOLocationID                      int32
payment_type                      int64
fare_amount                     float64


# 3. Feature Relevance Analysis for Hourly Taxi Demand Prediction

Which columns are useful for predicting demand (trip count per zone per hour)?

In [30]:
# Extract temporal features from pickup datetime to understand demand patterns
df['pickup_hour'] = df['tpep_pickup_datetime'].dt.hour
df['pickup_day_of_week'] = df['tpep_pickup_datetime'].dt.dayofweek
df['pickup_date'] = df['tpep_pickup_datetime'].dt.date

print("\n" + "=" * 80)
print("FEATURE RELEVANCE FOR HOURLY DEMAND PREDICTION")
print("=" * 80)

print("\n### KEY FEATURES ###")
print("\nHIGHLY RELEVANT Features:")
print("-" * 80)

print("\n1. PULocationID (Pickup Location):")
print("   - CRITICAL: This is our prediction target at the zone level")
print("   - Used to aggregate trip count per zone per hour")
print(f"   - {df['PULocationID'].nunique()} unique zones")

print("\n2. tpep_pickup_datetime (Pickup Timestamp):")
print("   - CRITICAL: Enables hourly aggregation (our target granularity)")
print("   - Enables extraction of temporal patterns (hour of day, day of week)")
print(f"   - Time range: {df['tpep_pickup_datetime'].min()} to {df['tpep_pickup_datetime'].max()}")

print("\n3. Temporal features (derived from pickup_datetime):")
print(f"   - Pickup hour (0-23): {sorted(df['pickup_hour'].unique())}")
print(f"   - Day of week (0=Mon, 6=Sun): {sorted(df['pickup_day_of_week'].unique())}")
print("   - These capture demand patterns (rush hours, weekday vs weekend)")

print("\n\nLIMITED/REDUCED RELEVANCE Features:")
print("-" * 80)

print("\n1. Trip-level features (not zone-level predictors):")
print("   - tpep_dropoff_datetime, DOLocationID, trip_distance, fare_amount, etc.")
print("   - These describe individual trips but don't directly predict zone demand")
print("   - Could be used for external features (e.g., events at dropoff zones)")

print("\n2. Potentially low-signal features:")
print("   - VendorID: Likely doesn't affect demand patterns")
print("   - payment_type, tip_amount: Describe payment, not demand generation")
print("   - trip_distance, fare_amount: Correlate with demand but aren't predictors")

# Demonstrate aggregation to hourly demand
print("\n" + "=" * 80)
print("EXAMPLE: Aggregating to Hourly Demand Target")
print("=" * 80)

# Create hourly demand by zone
df['hour'] = df['tpep_pickup_datetime'].dt.floor('h')  # Round to hour
hourly_demand = df.groupby(['hour', 'PULocationID']).size().reset_index(name='trip_count')

print(f"\nOriginal data: {len(df)} individual trips")
print(f"After hourly aggregation: {len(hourly_demand)} zone-hour combinations")
print(f"\nSample of hourly demand data:")
print(hourly_demand.head(10).to_string())

print(f"\nDemand statistics:")
print(f"  - Average trips per zone per hour: {hourly_demand['trip_count'].mean():.2f}")
print(f"  - Min: {hourly_demand['trip_count'].min()}, Max: {hourly_demand['trip_count'].max()}")
print(f"  - Std Dev: {hourly_demand['trip_count'].std():.2f}")

# Show which zones have highest demand
print(f"\nTop 10 zones by total trip count:")
zone_demand = df.groupby('PULocationID').size().sort_values(ascending=False)
print(zone_demand.head(10).to_string())


FEATURE RELEVANCE FOR HOURLY DEMAND PREDICTION

### KEY FEATURES ###

HIGHLY RELEVANT Features:
--------------------------------------------------------------------------------

1. PULocationID (Pickup Location):
   - CRITICAL: This is our prediction target at the zone level
   - Used to aggregate trip count per zone per hour
   - 261 unique zones

2. tpep_pickup_datetime (Pickup Timestamp):
   - CRITICAL: Enables hourly aggregation (our target granularity)
   - Enables extraction of temporal patterns (hour of day, day of week)
   - Time range: 2024-12-31 20:47:55 to 2025-02-01 00:00:44

3. Temporal features (derived from pickup_datetime):
   - Pickup hour (0-23): [np.int32(0), np.int32(1), np.int32(2), np.int32(3), np.int32(4), np.int32(5), np.int32(6), np.int32(7), np.int32(8), np.int32(9), np.int32(10), np.int32(11), np.int32(12), np.int32(13), np.int32(14), np.int32(15), np.int32(16), np.int32(17), np.int32(18), np.int32(19), np.int32(20), np.int32(21), np.int32(22), np.int32(23)]

# 4. Summary & Key Takeaways

In [31]:
print("\n" + "=" * 80)
print("SUMMARY: INSIGHTS FOR WEEK 2 STRUCTURED EDA")
print("=" * 80)

print("\n📊 DATASET SHAPE:")
print(f"   - {df.shape[0]:,} individual trips")
print(f"   - {df.shape[1]} fields per trip")
print(f"   - Data covers: {df['tpep_pickup_datetime'].min().date()} to {df['tpep_pickup_datetime'].max().date()}")

print("\n⚠️  EXPECTED DATA QUALITY CHECKS FOR WEEK 2:")
print("   - Missing values in optional fields (tip_amount, improvement_surcharge, etc.)")
print("   - Negative or zero values in fare, distance, passenger_count")
print("   - Trip duration anomalies (dropoff before pickup)")
print("   - Invalid location IDs (0 or out-of-range zones)")
print("   - Duplicates or exact repeated records")
print("   - Outliers in trip distance and fare amounts")

print("\n🎯 FEATURES FOR PREDICTING HOURLY DEMAND:")
print("   PRIMARY:")
print("   ✓ PULocationID - directly used to aggregate demand by zone")
print("   ✓ tpep_pickup_datetime - enables hourly binning and temporal features")
print("\n   SUPPORTING (for eventual predictive models):")
print("   ✓ Temporal patterns - hour of day, day of week (from pickup time)")
print("   ✓ Historical demand patterns - aggregate past demand by zone+hour")
print("   ✓ External factors - holidays, events, weather (not in this dataset)")
print("\n   NOT USEFUL FOR DIRECT PREDICTION:")
print("   ✗ Trip characteristics (distance, fare, passengers) - describe trips, not demand")
print("   ✗ Dropoff location - describes where trips end, not where they start")
print("   ✗ Payment/vendor info - not predictive of trip generation")


SUMMARY: INSIGHTS FOR WEEK 2 STRUCTURED EDA

📊 DATASET SHAPE:
   - 3,475,226 individual trips
   - 25 fields per trip
   - Data covers: 2024-12-31 to 2025-02-01

⚠️  EXPECTED DATA QUALITY CHECKS FOR WEEK 2:
   - Missing values in optional fields (tip_amount, improvement_surcharge, etc.)
   - Negative or zero values in fare, distance, passenger_count
   - Trip duration anomalies (dropoff before pickup)
   - Invalid location IDs (0 or out-of-range zones)
   - Duplicates or exact repeated records
   - Outliers in trip distance and fare amounts

🎯 FEATURES FOR PREDICTING HOURLY DEMAND:
   PRIMARY:
   ✓ PULocationID - directly used to aggregate demand by zone
   ✓ tpep_pickup_datetime - enables hourly binning and temporal features

   SUPPORTING (for eventual predictive models):
   ✓ Temporal patterns - hour of day, day of week (from pickup time)
   ✓ Historical demand patterns - aggregate past demand by zone+hour
   ✓ External factors - holidays, events, weather (not in this dataset)

   

================================================================================
SUMMARY: INSIGHTS FOR WEEK 2 STRUCTURED EDA
================================================================================

📊 DATASET SHAPE:
   - 3,475,226 individual trips
   - 25 fields per trip
   - Data covers: 2024-12-31 to 2025-02-01

⚠️  EXPECTED DATA QUALITY CHECKS FOR WEEK 2:
   - Missing values in optional fields (tip_amount, improvement_surcharge, etc.)
   - Negative or zero values in fare, distance, passenger_count
   - Trip duration anomalies (dropoff before pickup)
   - Invalid location IDs (0 or out-of-range zones)
   - Duplicates or exact repeated records
   - Outliers in trip distance and fare amounts

🎯 FEATURES FOR PREDICTING HOURLY DEMAND:
   PRIMARY:
   ✓ PULocationID - directly used to aggregate demand by zone
   ✓ tpep_pickup_datetime - enables hourly binning and temporal features

   SUPPORTING (for eventual predictive models):
   ✓ Temporal patterns - hour of day, day of week (from pickup time)
...
   NOT USEFUL FOR DIRECT PREDICTION:
   ✗ Trip characteristics (distance, fare, passengers) - describe trips, not demand
   ✗ Dropoff location - describes where trips end, not where they start
   ✗ Payment/vendor info - not predictive of trip generation